### ISSUE TITLE
Max pooling behavior for nan values

### ISSUE THREAD

**Issue by ClementPinard** [pytorch/pytorch#7645]
## Issue description

max pooling functions are not consistent with max functions.
Below an example, every max pooling (be it 1d, 2d or 3d, adaptive or not) acts the same, on cpu or on cuda.

Essentially, there are two fondamental differences :
 - max pooling of all `nan` values is `-THINF` while for `max` it's `nan`
 - max pooling of nan and valid values is valid values, which means `nan`s get ignored, while for `max`, as soon as there is a `nan` value, the result is `nan`.

More generally, choosing explicetely how to deal with `nan` as in numpy ([e.g.](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.nanmax.html)) could be a solution, but maybe this is related to CuDNN's max pooling ?

## Code example

```python
a= torch.full((1,1,2), float('nan'))
print(a.max())  # tensor(nan)
print(F.adaptive_max_pool1d(a, 1)) # tensor([[[-3.4028e+38]]])

a.requires_grad=True
a.max().backward()
print(a.grad) # tensor([[[ 0., 0.]]])
F.adaptive_max_pool1d(a, 1).backward()
print(a.grad) # tensor([[[ 0., 0.]]])

with torch.no_grad():
    a[0,0,0] = 0

a.max().backward()
print(a.grad) # tensor([[[ 0., 0.]]])
F.adaptive_max_pool1d(a, 1).backward()
print(a.grad) # tensor([[[ 1., 0.]]])
```

## System Info

Built from latest sources (as of 05/17)
PyTorch version: 0.5.0a0+331a04d
Is debug build: No
CUDA used to build PyTorch: 9.1.85

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: Quadro M1000M
Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[conda] magma-cuda91              2.3.0                         1    pytorch
[conda] torch                     0.5.0a0+331a04d           <pip>
[conda] torch                     0.3.1b0+2b47480           <pip>
[conda] torch                     0.5.0a0+4251e38           <pip>

**Comment by t-vi** [comment]
Hi, @ClementPinard . Thank you for providing the example. I'll take a look.

**Comment by t-vi** [comment]
Hmhm. I seem to get a double free in the `F.adaptive_max_pool1d(a, 1).backward()`...

**Comment by t-vi** [comment]
Another question: Do we really want to provide an option to ignore NaN? My view is that "if you have nan in your net, you're screwed", so I would just return NaN there, personally.

**Comment by ClementPinard** [comment]
My practical usecase is for KITTI groundtruth depthmaps and FlowMaps which are 2D sparse arrays.
For an algorithm that use FlowNet-like architecture that outputs predictions at multiple scale levels, we can either compare predictions to downscaled GT or upscaled predictions to GT.

The first being obviously less computationally expensive, the ignore NaN would help downscaling such sparse 2D maps.

For the moment we do something I find ugly, you can see it  [here](https://github.com/ClementPinard/FlowNetPytorch/blob/master/multiscaleloss.py#L19)
Essentially it takes the map and construct two maps of positive and negative values which are then maxpooled and added back together.

I am actually open for a strict "no-nan" policy on pooling functions, but in that case better enforce it before someone writes a code that tries to benefit from maxpooling ignore `nan`s feature/bug

**Comment by t-vi** [comment]
Personally, I think it is more sane have NaN -> NaN in the pooling and offer a parametrizable `nan_to_num` (where you get to pick the values) that does `torch.where(torch.isnan(x), torch.tensor(float('-inf')), x)` or so.

**Comment by t-vi** [comment]
So the proposed fix does NaN -> NaN similar to max. I didn't try to fix gradients for the nan case. This would involve keeping the values and I don't think that is worth it (in particular because I would not expect the pooling layer to be last, and otherwise we'd probably get NaN as grad_out).
If you are reasonably happy with it, I'd move it to a PR.

**Comment by ClementPinard** [comment]
+1 for NaN -> NaN  _"abyssus abyssum invocat"_ 
We could also add an optional mask which would be a ByteTensor of the same size, specifying whether or not the considered pixel is used for the pooling, that could be used for any kind of pulling. An `ignore_nan` functionality would then be to provide the mask `input != input`

**Comment by ClementPinard** [comment]
I could work with that.
to continue to be picky, the potential drawback is that it won't work for other pooling methods such as average pooling or median pooling. The problem here is that the max operation inherently ignores nonmax values which can be leveraged for "ignore some pixels" operation, but it has a "non universal" feel to it since it woun't be as easy for other kinds of 2D operations.

**Comment by ClementPinard** [comment]
To be clear, the fix is good for me, but I figured a related discussion on selective pooling (whether to ditch NaN values or anything you want to ignore) could happen (maybe not on this issue ?)

It would be convenient for my use case, but maybe for others too.

**Comment by soumith** [comment]
fixed via https://github.com/pytorch/pytorch/pull/7670
